### ISSUE TITLE
Align draw time to the screen refresh rate

### ISSUE THREAD

**Issue by HybridDog** [minetest/minetest#10283]
## Problem

Let's assume that the screen has a fixed 60 Hz frame rate.
As far as I know, if there is movement and the measured time between two adjacent images is not always 1/60 s but slightly deviates from this time (i.e. if there is jittering), Minetest would do the movement for those measured time differences and not 1/60 s delta time between the images. Then those images are aligned to the 60 Hz screen frame rate because the screen has this fixed refresh rate. The jittering could make motion in Minetest look a bit non-smooth.
This behaviour can be tested with a rotating object: If the screen has a 60 Hz rate, the object has 60 rotations per second and the draw time is correctly aligned to the screen's frame rate, the object should look like it does not rotate at all (temporal aliasing).

## Solutions

It is possible to align the delta time to 1/60 s (for a 60 Hz screen) in the code. This could make the game look more smooth, but has the disadvantage that visual jumps can occur when frames need to be skipped due to lag.
A setting could be added which en- or disables this aligning.

## Additional context

I'm not an expert in this topic and I don't know if the current Minetest already does something against jittering.

**Comment by paramat** [comment]
Like your previous issue, useful speculation about a possible problem, but no evidence of an actual problem yet, so no need to do anything yet.

**Comment by SmallJoker** [comment]
This already happens.
https://github.com/minetest/minetest/blob/abfea69e5f68c0f2c946bfcd4444f8cb32e781cf/src/client/game.cpp#L3980

**Comment by HybridDog** [comment]
> This already happens.

I don't see the code against jittering.
#
I have tried to implement this myself:
```diff
diff --git a/src/client/game.cpp b/src/client/game.cpp
index 0d3a0ca15..14ab9537e 100644
--- a/src/client/game.cpp
+++ b/src/client/game.cpp
@@ -1110,6 +1110,11 @@ void Game::run()
 		//  + Sleep time until the wanted FPS are reached
 		limitFps(&draw_times, &dtime);
 
+		static f32 dtime_rest = 0;
+		dtime += dtime_rest;
+		dtime_rest = fmodf(dtime, 1.0 / 60.0);
+		dtime -= dtime_rest;
+
 		// Prepare render data for next iteration
 
 		updateStats(&stats, draw_times, dtime);
```
This naive jittering reduction causes lots of frame skips for me. 
Maybe it is possible to somehow select a median of the last five dtime values so that dtime has low jittering.

**Comment by hecktest** [comment]
Doesn't vertical sync already do what you're looking for?

**Comment by HybridDog** [comment]
> Doesn't vertical sync already do what you're looking for?

I don't think that vsync avoids jittering. 
The problem is that `dtime` is not always `1/60 s ≈ 16.67 ms` but it can vary a bit between each frame, e.g. it could be `17.0 ms` or `16.4 ms` or some other value close to `1/60 s` between a pair of frames. On the other hand, the screen always waits `1/60 s` between displaying each frame. When a linear movement happens in the game, for example when the player flies with constant velocity, these small differences caused by the jittering may be visible to the player because, as far as I know, the human vision is sensitive to linear motions.
As I have already mentioned earlier, I'm no expert in this topic, so I don't know if it is important to avoid visual jittering.

**Comment by hecktest** [comment]
The jittering comes from the performance itself being jittery, and from the fact that we have to use the last frame's timing information. This may feel choppy if unstable, but it's also very accurate, the total distance travelled is approximated quite closely.

You could apply a low pass filter to delta time, but this would introduce inaccuracy into movement, making it partially framerate dependent, and probably no less jittery. It's not a good idea to use filtered delta time for anything where accuracy is desired.

Maybe this crude picture can explain. (or maybe not)
![deltat](https://user-images.githubusercontent.com/42101236/90763900-c940a080-e2e7-11ea-820d-c391904317ab.png)

The correct way to reduce jitter is to improve the engine's overall performance so that stutter does not happen in the first place.

**Comment by Hiradur** [comment]
I remember reading an interesting article related to this subject some years ago. [Here it is.](https://medium.com/@alen.ladavac/the-elusive-frame-timing-168f899aec92)
This article also mentions that simply making the engine assume that it is running at 60 fps can fix the issue when it actually is running at a constant 60 fps but causes problem when it is not. So it's probably not an universal solution due to different hardware setups, compute-intensive mods and so on...

I also remembered that the Yamagi Quake 2 engine source port advertises 
> Highly precise timings. For example, on a flat panel display with a 59.95 hertz refresh rate and vertical synchronisation (vsync) enabled Yamagi Quake II is running at exactly 59.95 frames per second.

[quoted from offical page](https://www.yamagi.org/quake2/)

I don't know if this is just a fancy way of describing vsync but playing Quake 2 with the Yamagi Quake II engine with vsync enabled was probably among the smoothest gaming experiences I had on my 60 Hz monitor so far.

Perhaps these source can serve as an inspiration.

**Comment by HybridDog** [comment]
> When a linear movement happens in the game, for example when the player flies with constant velocity, these small differences caused by the jittering may be visible to the player because, as far as I know, the human vision is sensitive to linear motions.

By now I think these differences may be difficult to notice on most displays because most displays are probably sample-and-hold displays. 
On a display which shows each frame only for a very short time and is black otherwise (e.g. with backlight strobing, black frame insertion, etc.), moving images look sharp and the jittering may be much more noticeable.

[This webpage](https://www.testufo.com/photo#photo=quebec.jpg) has a simple example where motion blur occurs. If the movement is paused, I can clearly read the "Q-52" on my display, otherwise it looks blurry.
[This video](https://www.youtube.com/watch?v=vNb3X1AM6uI) has a good explanation of the blurring effect in my opinion.

#

I have a new idea to align the frame rate. Assume that the display has a fixed refresh rate, the maximum frame rate in Minetest is $m$ and a frame is rendered such that it calculates the $i$-th frame for $t_{i,1}$ s and then sleeps $t_{i,2}$ s to limit the FPS.
* If the system is not fast enough to produce $m$ frames per second, we have $t_{i,1} \geq \frac{1}{m}$. In this case, we do not sleep, i.e. $t_{i,2} = 0$, we assume there is no noise, i.e. $n_i = 0$, and we use $t_{i,1} + n_{i-1}$ as dtime, i.e. we do not try to align dtime to the refresh rate $m$ and get rid of the noise from the previous frame.
* If the system is fast, i.e. $t_{i,1} < min(\frac{1}{m} - n_{i-1}, \frac{1}{m})$, we have $t_{i,1} + t_{i,2} = \frac{1}{m} + n_i$, where $n_i$ is the jittering noise. $n_i$ exists, for example, because we cannot sleep exactly $t_{i,2}$ seconds in practice. In this case, we set the sleep time to $t_{i,2} = \frac{1}{m} - t_{i,1} - n_{i-1}$ and the dtime value to $\frac{1}{m}$, i.e. we adjust the sleep time to get rid of previous jittering noise. 
* If, in the (probably) rare case, the system is on the brink of being too slow, we have $t_{i,1} < \frac{1}{m} < t_{i,1} + n_{i-1}$. In this case, we proceed like in the first case, i.e. $t_{i,2} = n_i = 0$ and we use $t_{i,1} + n_{i-1}$ as dtime. 

This method has the following properties (if I didn't make a mistake):
* The method may only work if we control the sleep times, so with vsync we may need a different approach.
* If the system is always too slow to reach the maximum frame rate, the behaviour is like in current Minetest because the noises $n_i$ are always zero.
* If the system is always calculating the frames fast, the dtime value is fixed to $\frac{1}{m}$ and the noise of the current frame is moved into the sleep time of the next frame. Here I assume that the noise $n_i$ does not have outliers, which I think is likely because the noise is due to sleep time inaccuracies.
* If the system is almost always calculating the frames fast but now and then an outlier occurs in the frame calculation time, the dtime value when an outlier occurs is almost the same as in current Minetest since the added noise can usually be neglected.

**Comment by HybridDog** [comment]
> The correct way to reduce jitter is to improve the engine's overall performance so that stutter does not happen in the first place.

No, according to [this page](https://raphlinus.github.io/ui/graphics/gpu/2021/10/22/swapchain-frame-pacing.html), OpenGL could wait for the display between handling the input and p

**Comment by HybridDog** [comment]
Now there's VK_EXT_present_timing, which is a major milestone in bringing time-based frame pacing to the Vulkan ecosystem.
Maybe we can use this together with OpenGL to fix the jittering problem.
https://www.khronos.org/blog/vk-ext-present-timing-the-journey-to-state-of-the-art-frame-pacing-in-vulkan
